### PR TITLE
Add CollectionOutputFormat for local executor

### PR DESCRIPTION
**Pull request by qmlmoon**
implement issue #602

### CODE REVIEW

File: stratosphere-java/src/main/java/eu/stratosphere/api/java/io/CollectionOutputFormat.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	public void configure(Configuration parameters) {
+	}
+
+
+	@Override
+	public void open(int taskNumber, int numTasks) throws IOException {
+		try {
+			this.taskResult = (Collection<T>) resultWrapper.getClass().newInstance();
+		} catch (InstantiationException e) {
+			e.printStackTrace();
```
Comment (rmetzger) [inline stratosphere-java/src/main/java/eu/stratosphere/api/java/io/CollectionOutputFormat.java]: Can you re-throw the exceptions as IOExceptions?

File: stratosphere-java/src/main/java/eu/stratosphere/api/java/DataSet.java [hunk 2]
```diff
@@ -191,7 +195,10 @@ public ExecutionEnvironment getExecutionEnvironment() {
 	public void writeAsText(String filePath) {
 		output(new TextOutputFormat<T>(new Path(filePath)));
 	}
-	
+
+	public <X extends Serializable> void writeAsCollection(Collection<X> out) {
```
Comment (fhueske) [inline stratosphere-java/src/main/java/eu/stratosphere/api/java/DataSet.java]: Shouldn't this be `writeAsCollection(Collection<T> out)` to ensure type safety of the collection?

Comment (qmlmoon) [inline stratosphere-java/src/main/java/eu/stratosphere/api/java/DataSet.java]: Yeah, it should be. But to copy the record object in `writeRecord(T record)`, I used `SerializationUtil.clone`

Comment (fhueske) [inline stratosphere-java/src/main/java/eu/stratosphere/api/java/DataSet.java]: You can get a type preserving serializer from TypeInformation of DataSet and set it in the constructor of ColloctionOutputFormat.

File: stratosphere-java/src/main/java/eu/stratosphere/api/java/io/CollectionOutputFormat.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.Collection;
+
+/**
+ *  An output format that writes record into collection
+ */
+public class CollectionOutputFormat<T> implements OutputFormat<T> {
+
+	private static final long serialVersionUID = 1L;
+
+	private static Object resultWrapper;   //since generic collection is not allowed to be static
```
Comment (StephanEwen) [inline stratosphere-java/src/main/java/eu/stratosphere/api/java/io/CollectionOutputFormat.java]: I think you can make this a `Collection<?>`.

File: stratosphere-java/src/main/java/eu/stratosphere/api/java/io/CollectionOutputFormat.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	public void writeRecord(T record) throws IOException {
+		T recordCopy = this.typeSerializer.createInstance();
+		this.typeSerializer.copy(record, recordCopy);
+		this.taskResult.add(recordCopy);
+	}
+
+
+	@Override
+	public void close() throws IOException {
+		Collection<T> result = (Collection<T>) this.resultWrapper;
```
Comment (StephanEwen) [inline stratosphere-java/src/main/java/eu/stratosphere/api/java/io/CollectionOutputFormat.java]: I think this is problematic in the parallel case. If many parallel tasks are in `close()`at the same time, they will concurrently add elements, which causes a `ConcurrentModificationException`.

You need to synchronize around the target list here.

**Comment by StephanEwen** [comment]
It is a good start. There are two things I would like to improve:
- I don't think we should have a dedicated method for this on `DataSet`. because this will only work for local testing. I would prefer to use it only through `result.output(new CollectionOutputFormat(list))`.
- This will not work for jobs with multiple sinks, because you have only one static reference. If you had two CollectionOutputFormats in the same program, both would get the union of the results. I think you can work around that by giving each CollectionOutputFormat a generated id and have a static map of collections, rather than only one static field.

File: stratosphere-java/src/main/java/eu/stratosphere/api/java/io/CollectionOutputFormat.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+
+	@Override
+	public void configure(Configuration parameters) {
+	}
+
+
+	@Override
+	public void open(int taskNumber, int numTasks) throws IOException {
+		try {
```
Comment (StephanEwen) [inline stratosphere-java/src/main/java/eu/stratosphere/api/java/io/CollectionOutputFormat.java]: I think you can just create a new `ArrayList` here. The collection in which you gather the elements needs not be of the same type as the collection into which you out the elements at the end. That spares you the reflection instantiation, which may fail if no public nullary constructor exists.

**Comment by qmlmoon** [comment]
`CollectionOutputFormat` needs a `TypeSerializer` from `DataSet` to make a copy of record. To remove the method on `DataSet`, I would do something like `CollectionOutputFormat<T extends Tuple>` to use `SerializationUtil.clone()`, which forces the element of the collection to be a tuple. But it's not good for type safty as Fabian pointed.

**Comment by StephanEwen** [comment]
I see. But I don't think the workaround can be to add a method to data set for every possible output format.
Let's add an interface called `TypeConfigurable` or so. Then the output() function can call a method on corresponding output formats and pass the type.

**Comment by qmlmoon** [comment]
What about to use the collection like `result.output(new CollectionOutputFormat(list, result.getType().createSerializer))` ,just specify the serialiser manually

File: stratosphere-java/src/main/java/eu/stratosphere/api/java/io/CollectionOutputFormat.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+	@Override
+	public void open(int taskNumber, int numTasks) throws IOException {
+		this.taskResult = new ArrayList<T>();
+	}
+
+	@Override
+	public void writeRecord(T record) throws IOException {
+		T recordCopy = this.typeSerializer.createInstance();
+		this.typeSerializer.copy(record, recordCopy);
```
Comment (StephanEwen) [inline stratosphere-java/src/main/java/eu/stratosphere/api/java/io/CollectionOutputFormat.java]: I think this needs to be `recordCopy = this.typeSerializer.copy(record, recordCopy);`

File: stratosphere-java/src/main/java/eu/stratosphere/api/java/io/CollectionOutputFormat.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+	public void close() throws IOException {
+		synchronized (this.resultWrapper) {
+			Collection<T> result = (Collection<T>) this.resultWrapper.get(this.id);
+			result.addAll(this.taskResult);
+		}
+	}
+
+	@Override
+	public void setInputType(TypeInformation<?> type) {
+		this.typeSerializer = (TypeSerializer<T>)type.createSerializer();
```
Comment (StephanEwen) [inline stratosphere-java/src/main/java/eu/stratosphere/api/java/io/CollectionOutputFormat.java]: Can you suppress the "unchecked" warnings here?

**Comment by StephanEwen** [comment]
Looks nice to me :-)

I think we should rename it, just to make sure that it is obvious that this thing is only intended for local executor use. Maybe call it `LocalCollectionOutputFormat` or so.

**Comment by fhueske** [comment]
Good to merge, thanks! :+1:

**Comment by fhueske** [comment]
Will merge this now.

**Pull request closed without merging**